### PR TITLE
Add basic vision

### PR DESCRIPTION
**Pull request by ciaran-regan-ie**
Adds a vision implementation in `/examples/vision`.

This vision example is very simple, and does not do any inference.

### CODE REVIEW

**Comment by krmr73** [comment]
@ciaran-regan-ie 
I tried `examples/vision/example.jl` and got the following error. It appears that the changes in `PDDLViz.jl` are not reflected. Do you know what I should do?
```
➜  Plinf.jl git:(ciaran/add-basic-vision) julia examples/vision/example.jl
Registering PDDL array theory
Loading domain
Loading problem
Initializing state
Initial state: GenericState(Set(Compound[onion(onion1), carrot(carrot1)]), Set{Term}(), Dict{Symbol, Any}(:walls => Bool[0 0 0 0 0 0 0 0 0 0; 0 1 1 1 0 0 1 1 1 0; 0 1 0 0 0 1 0 0 1 0; 0 1 0 1 0 1 1 0 1 0; 0 0 0 1 0 0 0 0 0 0; 0 1 1 0 0 1 1 1 0 1; 0 1 0 0 1 0 0 1 0 1; 0 0 0 1 0 1 0 0 0 0; 0 1 0 0 0 0 1 1 0 1; 0 0 0 0 0 0 0 0 0 0], :xloc => Dict{Any, Any}((:onion1,) => 5, (:carrot1,) => 8), :yloc => Dict{Any, Any}((:onion1,) => 5, (:carrot1,) => 8), :viewdistance => Bool[0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0], :xpos => 1, :ypos => 1))
Constructing goal specification
Compiling domain
Defining renderer
ERROR: LoadError: MethodError: no method matching GridworldRenderer(; resolution=(600, 700), agent_renderer=var"#31#36"(), obj_renderers=Dict{Symbol, Function}(:carrot => var"#32#37"(), :onion => var"#33#38"()), show_inventory=true, inventory_fns=[var"#34#39"()], inventory_types=[:item], show_vision=true, vision_fns=[var"#35#40"()], vision_types=[:item])
Closest candidates are:
  GridworldRenderer(; resolution, grid_fluents, grid_colors, has_agent, get_agent_x, get_agent_y, get_obj_x, get_obj_y, agent_renderer, obj_renderers, obj_type_z_order, locations, show_inventory, inventory_fns, inventory_types, inventory_labels, inventory_labelsize, state_options, trajectory_options, anim_options) at util.jl:493 got unsupported keyword arguments "show_vision", "vision_fns", "vision_types"
  GridworldRenderer(::Tuple{Int64, Int64}, ::Vector{Term}, ::Vector, ::Bool, ::Function, ::Function, ::Function, ::Function, ::Function, ::Dict{Symbol, Function}, ::Vector{Symbol}, ::Vector{Tuple}, ::Bool, ::Vector{Function}, ::Vector{Symbol}, ::Vector{String}, ::Real, ::Dict{Symbol, Any}, ::Dict{Symbol, Any}, ::Dict{Symbol, Any}) at ~/.julia/packages/PDDLViz/icukg/src/renderers/gridworld/gridworld.jl:13 got unsupported keyword arguments "resolution", "agent_renderer", "obj_renderers", "show_inventory", "inventory_fns", "inventory_types", "show_vision", "vision_fns", "vision_types"
  GridworldRenderer(::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any) at ~/.julia/packages/PDDLViz/icukg/src/renderers/gridworld/gridworld.jl:13 got unsupported keyword arguments "resolution", "agent_renderer", "obj_renderers", "show_inventory", "inventory_fns", "inventory_types", "show_vision", "vision_fns", "vision_types"
Stacktrace:
 [1] kwerr(kw::NamedTuple{(:resolution, :agent_renderer, :obj_renderers, :show_inventory, :inventory_fns, :inventory_types, :show_vision, :vision_fns, :vision_types), Tuple{Tuple{Int64, Int64}, var"#31#36", Dict{Symbol, Function}, Bool, Vector{var"#34#39"}, Vector{Symbol}, Bool, Vector{var"#35#40"}, Vector{Symbol}}}, args::Type)
   @ Base ./error.jl:165
 [2] top-level scope
   @ ~/Documents/research/mixi/connectsphere/Plinf.jl/examples/vision/example.jl:33
in expression starting at /Users/nanami/Documents/research/mixi/connectsphere/Plinf.jl/examples/vision/example.jl:33
```

The following command was executed.
```
pkg> add https://github.com/ConnectSphereInc/PDDLViz.jl#ciaran/add-vision
```

**Comment by ciaran-regan-ie** [comment]
@krmr73 Thank you for finding that bug! It seems like I was running a different version of PDDLViz cached locally. 

I have push new commits to both repos. Can you try again please?

File: examples/vision/example.jl
Comment (krmr73) [inline examples/vision/example.jl:1]: Thank you for fixing the bug! I could run this!

Comment (krmr73) [inline examples/vision/example.jl:1]: Thank you for fixing the bug! I could run this!

File: examples/vision/example.jl [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+# Register PDDL array theory
+println("Registering PDDL array theory")
+PDDL.Arrays.register!()
+
+# Load domain and problem
+println("Loading domain")
+domain = load_domain(joinpath(@__DIR__, "domain.pddl"))
+println("Loading problem")
+problem = load_problem(joinpath(@__DIR__, "problems", "problem-2.pddl"))
```
Comment (krmr73) [inline examples/vision/example.jl:16]: I also tried "problem-1.pddl" but it wasn't working. Is this expected?

```
Registering PDDL array theory
Loading domain
Loading problem
Initializing state
Initial state: GenericState(Set(Compound[onion(onion1), carrot(carrot1)]), Set{Term}(), Dict{Symbol, Any}(:walls => Bool[0 0 0 0; 0 1 1 0; 0 0 0 0; 0 0 0 0], :xloc => Dict{Any, Any}((:onion1,) => 3, (:carrot1,) => 4), :yloc => Dict{Any, Any}((:onion1,) => 3, (:carrot1,) => 4), :viewdistance => 2, :xpos => 1, :ypos => 1))
Constructing goal specification
Compiling domain
ERROR: LoadError: MethodError: Cannot `convert` an object of type Int64 to an object of type BitMatrix
Closest candidates are:
  convert(::Type{T}, ::LinearAlgebra.Factorization) where T<:AbstractArray at /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/LinearAlgebra/src/factorization.jl:58
  convert(::Type{<:BitArray}, ::AbstractArray) at bitarray.jl:580
  convert(::Type{T}, ::T) where T<:AbstractArray at abstractarray.jl:16
  ...
Stacktrace:
 [1] setproperty!(x::PDDL.var"##CompiledVisionState#315", f::Symbol, v::Int64)
   @ Base ./Base.jl:39
 [2] set_fluent!(state::PDDL.var"##CompiledVisionState#315", val::Int64, term::Const)
   @ PDDL ~/.julia/packages/PDDL/rIUmS/src/compiler/state.jl:181
 [3] PDDL.var"##CompiledVisionState#315"(state::GenericState)
   @ PDDL ~/.julia/packages/PDDL/rIUmS/src/compiler/state.jl:89
 [4] top-level scope
   @ ~/.julia/packages/PDDL/rIUmS/src/compiler/state.jl:13
 [5] eval
   @ ./boot.jl:368 [inlined]
 [6] eval
   @ ~/.julia/packages/PDDL/rIUmS/src/PDDL.jl:1 [inlined]
 [7] compiled(domain::GenericDomain, state::GenericState)
   @ PDDL ~/.julia/packages/PDDL/rIUmS/src/compiler/compiler.jl:62
 [8] top-level scope
   @ ~/Documents/research/mixi/connectsphere/Plinf.jl/examples/vision/example.jl:27
in expression starting at /Users/nanami/Documents/research/mixi/connectsphere/Plinf.jl/examples/vision/example.jl:27
```

File: examples/vision/example.jl
Comment (krmr73) [inline examples/vision/example.jl:1]: What does the output state indicate? Especially the last such as `[8, 5], 1, [8, 5], 1` change with each step, what are these value?
```
Step 1: right
Current state: PDDL.var"##CompiledVisionState#315"(Bool[0, 0], Bool[0, 0], Bool[0, 0], Bool[0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0; 0 0 0 0 0 0 0 0 0 0], Bool[0 0 0 0 0 0 0 0 0 0; 0 1 1 1 0 0 1 1 1 0; 0 1 0 0 0 1 0 0 1 0; 0 1 0 1 0 1 1 0 1 0; 0 0 0 1 0 0 0 0 0 0; 0 1 1 0 0 1 1 1 0 1; 0 1 0 0 1 0 0 1 0 1; 0 0 0 1 0 1 0 0 0 0; 0 1 0 0 0 0 1 1 0 1; 0 0 0 0 0 0 0 0 0 0], [8, 5], 1, [8, 5], 1)
```

**Review by krmr73 — APPROVED** [review]
Great work!! I have a few questions, please check them.
I may ask again when I implement the search method.

File: examples/vision/example.jl (hunk 1 above)
Comment (ciaran-regan-ie) [reply examples/vision/example.jl:16]: You found another bug! Great work :D

It was a problem with a definition in the domain file not matching the problem file. I have removed the viewdistance parameter, so should be okay now

File: examples/vision/example.jl
Comment (ciaran-regan-ie) [reply examples/vision/example.jl:1]: Bool[0, 0], Bool[0, 0], Bool[0, 0] 
These indicate the states of the predicates `has`, `offgrid`, `visible` for each object (carrot or onion so 2d array). 0 for false and 1 for true

 Bool[0 0 0 0; 0 1 1 0; 0 0 0 0; 0 0 0 0]
 This is the bitmatrix depicting the layout of the walls
 
 [4, 3], 1, [4, 3], 1
Here 1 and 1 indicate the agents position: `xpos`, `ypos`
The 2d arrays are the xlocs and ylocs of the objects (carrot and onion so 2 objects/2d array)

Comment (ciaran-regan-ie) [reply examples/vision/example.jl:1]: image

File: examples/vision/example.jl (hunk 1 above)
Comment (krmr73) [reply examples/vision/example.jl:16]: I see. Thank you for fixing it so quickly!

**Pull request merged**